opal/datatype/opal_convertor.c

```diff
@@ -324,6 +324,215 @@ int32_t opal_convertor_unpack( opal_convertor_t* pConv,
+    convertor->pStack[1].disp = local_position - track;
+    convertor->pStack[0].disp = done_count * (pData->ub - pData->lb);
+
+    return 0;
+}
+
+int32_t
+opal_iovec_pack( opal_convertor_t *convertor,
+        struct iovec *out_iov,
+        uint32_t *out_size,
+        size_t *max_data )
+{
+    const opal_datatype_t *pData = convertor->pDesc;
+    struct iovec *iov = pData->iov;
+    char *dst = out_iov[0].iov_base,
+         *src = convertor->pBaseBuf + convertor->pStack[0].disp;
+    size_t track = *max_data, iov_track;
+    uint32_t i, iov_count = 0;
+
+    dst = out_iov[iov_count].iov_base;
+    iov_track = out_iov[iov_count].iov_len;
+
+    while( convertor->pStack[0].count ){
+        for( i = convertor->pStack[1].index; i < pData->iovcnt; i++ ) {
+            /* exceed chunk size */
+            if( track == 0 ){
+                convertor->pStack[1].index = i;
+                goto complete_pack;
+            }
+
+restart_pack:
+            if( iov_track < (iov[i].iov_len - convertor->pStack[1].disp) && iov_track < track ){
+                memcpy( dst,
+                        src + (ptrdiff_t)(iov[i].iov_base) + convertor->pStack[1].disp,
+                        iov_track);
+                convertor->pStack[1].disp += iov_track;
+                convertor->pStack[1].index = i;
+                track -= iov_track;
+
+                iov_count++;
+                if( iov_count == *out_size )
+                    goto complete_pack;
+
+                dst = out_iov[iov_count].iov_base;
+                iov_track = out_iov[iov_count].iov_len;
+
+                goto restart_pack;
+            }
+
+            if( track < (iov[i].iov_len - convertor->pStack[1].disp) ){
+                memcpy( dst,
+                        src + (ptrdiff_t)(iov[i].iov_base) + convertor->pStack[1].disp,
+                        track);
+
+                convertor->pStack[1].disp  += track;
+                convertor->pStack[1].index = i;
+                track = 0;
+
+                goto complete_pack;
+            }
+
+            /* or not */
+            memcpy( dst,
+                    src + (ptrdiff_t)(iov[i].iov_base) + convertor->pStack[1].disp,
+                    iov[i].iov_len - convertor->pStack[1].disp );
+
+            dst += iov[i].iov_len - convertor->pStack[1].disp;
+            iov_track -= iov[i].iov_len - convertor->pStack[1].disp;
+            track -= iov[i].iov_len - convertor->pStack[1].disp;
+            convertor->pStack[1].disp = 0;
+        }
+
+        convertor->pStack[0].disp += pData->ub - pData->lb;
+        convertor->pStack[1].index = 0;
+        convertor->pStack[0].count--;
+
+        src += pData->ub - pData->lb;
+    }
+
+complete_pack:
+    *max_data -= track;
+    convertor->bConverted += *max_data;
+
+    if( convertor->bConverted < convertor->local_size ){
+        return 0;
+    }
+
+    convertor->flags |= CONVERTOR_COMPLETED;
+    return 1;
+}
+
+int32_t
+opal_iovec_unpack( opal_convertor_t *convertor,
```

Comment:
Once you fix the pack, the unpack should be identical with the exception of the direction of the memcpy.